_source/_posts/2021-12-03-antipatterns-secrets-java.adoc

```diff
@@ -0,0 +1,149 @@
+---
+layout: blog_post
+title: "5 Anti-Patterns with Secrets in Java"
+author: brian-demers
+by: advocate
+communities: [security,java]
+description: "Learn how how to better manage and protect your passwords and API keys!"
+tags: [java, security, secrets, cryptography, tips]
+tweets:
+- "When was the last time you checked a password into source control? 🤦 Here are a few tips to avoid that and more‼️"
+- "API keys, passwords, and secrets. Learn a few common anti-patterns to avoid ⚠️"
+- "Secrets, passwords, and API keys oh my‼️ Make sure you are handling them correctly 🧑‍💻"
+image:
+type: awareness
+---
+
+:toc: macro
+:page-liquid:
+:experimental:
+
+Most applications use some secret or password: database connection info, API keys, OAuth client secrets, JWT keys, etc. Dealing with these secrets in your projects is always a chore, and it's commonly done wrong. In this post, I'll explain a few common problems and offer suggestions to help you avoid these issues.
+
+toc::[]
+
+== 1. Secrets checked into source control
+
+Checking a secret into source control is easy to do, and most of us have done it at one point, even if we don't want to admit it. Once a secret is in your source control, it's in there forever; it only takes https://github.com/search?q=removed+password&type=commits[a quick history search] to find them.
+
+If you find a secret in your source, remove it and revoke it. It's best if secrets never get checked in the first place, watch out for them in code reviews and scan your code bases to ensure this doesn't happen: https://www.gitguardian.com/[GitGuardian] and https://github.com/awslabs/git-secrets[git-secrets].
+
+Secrets often enter source trees by accident, usually because you temporarily typed one into a config file. When secrets are mixed in with other application properties, it makes them harder to spot or simply forgotten about. Keep secrets out of your source entirely by using an https://12factor.net/config[external configuration] provider like Hashicorp Vault, using environmental variables], or by keeping secrets in a separate file outside of your source tree (or one that is explicitly ignored in your `.gitignore` file.)
+
+TIP: Check out the https://github.com/paulschwarz/spring-dotenv[spring-dotenv] to use `.env` files with your Spring projects.
+
+
+== 2. Confusing encoding with encryption
+
+Many folks are confused about the difference between encoding and encryption; they are very distinct operations and cannot be interchanged.
+
+=== Encoding
+
+Data that is _encoded_ is transformed into another format, usually to make transferring between systems possible. An encoded value may or may not be human-readable, but it needs to be treated the same way as the original input source.  The following examples show a string encoded in two common encoding schemes URL encoding.
```

Comment:
```suggestion
Data that is _encoded_ is transformed into another format, usually to make transferring between systems possible. An encoded value may or may not be human-readable, but it needs to be treated the same way as the original input source.  The following examples show a string encoded in two common encoding schemes URL encoding.
```

I don't understand "URL encoding" in the closing sentence. Is there a typo? Or, does this work?: " The following examples show two common schemes for encoding a string."

Reply:
```suggestion
Data that is _encoded_ is transformed into another format, usually to make transferring between systems possible. An encoded value may or may not be human-readable, but it needs to be treated the same way as the original input source.  The following examples show a string encoded in two common encoding schemes: URL encoding and base64 encoding.
```